Add per-element unique op for CPU

theweiho opened the pull request:
Questions/possible future works:
* How to template-ize to extend support beyond LongTensor?
* How to check if autograd works (and if not, how to add explicit gradient)?
* CUDA support?

Testing command:
DEBUG=1 NO_CUDA=1 MACOSX_DEPLOYMENT_TARGET=10.9 CC=clang CXX=clang++ python setup.py build &&  DEBUG=1 NO_CUDA=1 MACOSX_DEPLOYMENT_TARGET=10.9 CC=clang CXX=clang++ python setup.py develop && python3 test/test_torch.py

Commands to preview generated documentations:
cd docs
pip install -r requirements.txt
make html

theweiho committed:
Initial commit for unique op

theweiho committed:
Working unique with test

theweiho committed:
Make inverse indices shape conform to input

theweiho committed:
flake8 whitespace removal

theweiho commented:
@soumith @goldsborough @jamesr66a 
pytorch/pytorch#2031
onnx/onnx#568

goldsborough commented on aten/src/ATen/native/Unique.cpp:
```diff
@@ -0,0 +1,45 @@
+  Tensor inverse_indices;
+  if (!return_inverse) {
+    inverse_indices = self.type().toScalarType(kLong).tensor({0});
+    
```
Nit: remove newline Actually we have a `.clang-format` here, so you can just auto-format it (e.g. using Nuclide)

goldsborough replied:
Also flip the cases again

theweiho replied:
Done

goldsborough commented on aten/src/ATen/native/Unique.cpp:
```diff
@@ -0,0 +1,45 @@
+    inverse_indices = self.type().toScalarType(kLong).tensor({0});
+    
+  } else {
+    inverse_indices = self.type().toScalarType(kLong).tensor(self.sizes());    
```
I think you want `at::empty_like(self)` or `at::zero_like(self)`. That will create a tensor with the same shape and type

theweiho replied:
For the inverse indices, though, I think it always make sense for them to be of type kLong regardless of what the input tensor type is (assuming support for non-LongTensor input is added) - I just want them to have the same shape as the input

goldsborough replied:
Then maybe `at::CPU(kLong).tensor(self.sizes())`

theweiho replied:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/README.md#can-it-handle-being-passed-variables specifically calls out to never use at::CPU to deal with inheritance issues - is that not relevant in this case?

goldsborough replied:
No you're right, I wasn't aware of this. Let's leave it then.

gchanan commented:
On your questions:
1) Look at the dispatch functions in Dispatch.h
2) You can write a test in test/test_autograd.py; if you need an explicit gradient you can add it in derivatives.yaml
3) You can use a dispatch: declaration in native_functions.yaml; see CUDA: examples in that file.

theweiho commented on aten/src/ATen/native/Unique.cpp:
```diff
@@ -0,0 +1,45 @@
+      self.data<int64_t>(), self.data<int64_t>() + self.numel());
+  Tensor output = self.type().tensor({static_cast<long long>(set.size())});
+  
+  if (!sorted) {
```
Done

theweiho commented on aten/src/ATen/native/Unique.cpp:
```diff
@@ -0,0 +1,45 @@
+// Returns unique elements of input tensor.
+
+#include <unordered_map>
```
Done

goldsborough commented:
Look great, had mostly nits.
For generic types, the way you'll want to do this is move all your logic into a separate template function (e.g. in an anonymous namespace above):
```
template<typename scalar_t>
void unique_op(at::Tensor ...) {
// use scalar_t instead of uint64_t here
}
```

then in the current `unique` function, something like:
```
AT_DISPATCH_ALL_TYPES(self.type(), "unique", [&] {
  unique_op<scalar_>t(tensors...);
});
```

See [here](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/EmbeddingBag.cu#L269) for an example

theweiho committed:
address review comment nits

SsnL commented on aten/src/ATen/native/Unique.cpp:
```diff
@@ -0,0 +1,44 @@
+    const Tensor& self, const bool sorted, const bool return_inverse) {
+  std::unordered_set<int64_t> set(
+      self.data<int64_t>(), self.data<int64_t>() + self.numel());
+  Tensor output = self.type().tensor({static_cast<long long>(set.size())});
```
I don't think we should use an `unordered_set` here. The `!sorted` case below has no guarantee that `set.begin()` and `set.end()` is ordered in a way. 

In fact, why do you need a set? Can't you just use torch's sort (if needed)? Then just iterate through once will be enough. Also it gives indices so computing `inverse_indices` should also be convenient.

goldsborough replied:
The `!sorted` case doesn't require ordering I think?
Also this set method is `O(N)` (with `O(N)` space) while sorting is `O(NlogN)` so slower complexity wise (if anyone cares)

SsnL replied:
Oh right my bad. I thought misunderstood `sorted` arg. I thought it was saying that the input is sorted.

SsnL commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -363,6 +363,9 @@
   variants: method
 
+- func: unique(IndexTensor self, bool sorted=false, bool return_inverse=false) -> (Tensor, Tensor)
+  variants: method
```
why is it only a method?

theweiho replied:
No reason - just wasn't sure what the convention for method vs function (or both) was. Added function

ssnl reviewed (changes requested):
This at its current state still needs more work:
1. Documentation at `_torch_docs.py` and `_tensor_docs.py`
2. Dispatch to other types can be in a later PR, but this will fail with cuda long tensor (probably a segfault). Can you add a cuda version as well?

theweiho committed:
Expose fn and add docs. Explicitly declare no gradients

theweiho committed:
Trial generic dispatch implementation

theweiho committed:
Add tests for generics

goldsborough commented:
I think you accidentally merged instead of rebasing

theweiho committed:
flake8 whitespace

theweiho commented:
Fixed merge. Documentations added and exposed unique as function as well. In order to do so, it turns out using `IndexTensor` in `native_functions.yaml` makes it so that the fn is not exposed as `torch.unique()`, but only as `torch.autograd.Variable.unique()` (thanks @gchanan!) - so changed to generic Tensor and also added dispatch to other types.

On further thought, realized that there's not really a sensible way to do gradients for this, so explicitly declared it as not implemented.

Regarding CUDA support, @SsnL - I pinged you, but it seems you're off the grid? In any case, I'd probably prefer to defer that - so is there a way to explicitly warn/catch CUDA usage for more graceful failure than segfault?

goldsborough commented on aten/src/ATen/native/Unique.cpp:
```diff
@@ -0,0 +1,67 @@
+#include <unordered_map>
+#include <unordered_set>
+
+#include <iostream>
```
Looks like this should be removed

theweiho replied:
Done

apaszke commented on test/test_torch.py:
```diff
@@ -5422,6 +5422,65 @@ def test_set_flush_denormal(self):
+            return_inverse=True,
+        )
+        self.assertEqual(torch.ByteTensor([7, 42, 128, 133]), byte_unique)
+        self.assertEqual(torch.LongTensor([3, 0, 0, 0, 1, 2]), byte_inverse)
```
It would be better to implement `unique` reference implementation in Python, and then have a list of examples that we loop over and compare outputs. Here's an example:
```python
def unique(tensor):
    values = sorted(set(tensor))
    value_map = {i: v for i, v in enumerate(values)}
    inverse = torch.LongTensor([value_map[v.item()] for v in values])
    return values, inverse
```

theweiho replied:
Hmm - for unit tests, I usually prefer to err on the side of verbosity and explicitness rather than doing things programmatically - since that makes it easier for people to see what the expected return values are and also serve as additional documentation for what we expect the function to return.

I could change to doing a reference implementation though, if you feel strongly about this.

apaszke commented on torch/_torch_docs.py:
```diff
@@ -4926,6 +4926,62 @@
+        - **output** (*Tensor*): the list of unique scalar elements
+        - **inverse_indices** (*Tensor*): the indices (same shape as input)
+            for where elements in the original input map to in the output
+            if ``return_inverse`` is ``True``; otherwise, an empty tensor.
```
Please format the docs according to the format we use everywhere else

theweiho replied:
This is following the format from `torch.eig` - can you clarify which one I should follow instead? In particular, I was looking for a format that returns a tuple of Tensor and clearly documents what each of the returned tensor means. The `Returns:` section seems to be inconsistently used - is that not part of the format?

fmassa commented:
@theweiho about gradients for `unique`, can't we consider it's gradient as similar to indexing only the first unique elements of a tensor?

theweiho committed:
Add basic CUDA error throwing and templateize set

theweiho commented on aten/src/ATen/native/Unique.cpp:
```diff
@@ -0,0 +1,67 @@
+  if (sorted) {
+    std::vector<scalar_t> vec(set.begin(), set.end());
+    std::sort(vec.begin(), vec.end());
+    std::copy(vec.begin(), vec.end(), output->data<scalar_t>());
```
Done - thanks for the suggestion!

theweiho commented on test/test_torch.py:
```diff
@@ -5422,6 +5422,65 @@ def test_set_flush_denormal(self):
+        self.assertEqual(empty_inverse, x_inverse)
+
+        x_unique, x_inverse = torch.autograd.Variable.unique(
+            x, sorted=True, return_inverse=True)
```
Fixed. Missed updating this call after I figured out how to expose torch.unique()

theweiho commented:
@SsnL - is there some way to use Travis CI (or something else) to test the CUDA part of the code you suggested?

@fmassa - that was the way we were considering doing, but not sure that totally makes sense. Attributing the gradient to the first unique element seems a bit arbitrary? (Considering all occurrences of that element "contributes" to the unique equally) I also wasn't sure what a use case for the unique gradient would be, so figured it may make sense to defer it until someone has concrete requirements.

colesbury commented on aten/src/ATen/native/Unique.cpp:
```diff
@@ -0,0 +1,65 @@
+    const bool return_inverse,
+    Tensor* output,
+    Tensor* inverse_indices) {
+  set_type<scalar_t> set(
```
This looks like it assumes that `self` is contiguous. Use Tensor.contiguous() either here or in unique_cpu.

theweiho replied:
Done

colesbury commented on aten/src/ATen/native/Unique.cpp:
```diff
@@ -0,0 +1,65 @@
+
+  if (sorted) {
+    AT_DISPATCH_ALL_TYPES(self.type(), "unique", [&] {
+      _unique_cpu_template<std::set, scalar_t>(
```
I believe `AT_DISPATCH_ALL_TYPES` can return a value. I think it will be a bit cleaner to make `output` and `inverse_indices` return values from `_unique_cpu_template` instead of output arguments.

theweiho replied:
Done

colesbury commented on aten/src/ATen/native/cuda/Unique.cu:
```diff
@@ -0,0 +1,17 @@
+  throw std::runtime_error(
+      "unique is currently CPU-only, and lacks CUDA support. "
+      "Pull requests welcome!");
+  return std::make_tuple(self.type().tensor({0}), self.type().tensor({0}));
```
No need for the return statement. It's unreachable.

theweiho replied:
Done

theweiho committed:
Explicit contiguous and AT_DISPATCH_ALL_TYPES return

theweiho committed:
Remove extraneous numpy conversion

apaszke commented on aten/src/ATen/native/Unique.cpp:
```diff
@@ -0,0 +1,58 @@
+    }
+    for (int i = 0; i < input.numel(); ++i) {
+      inverse_indices.data<int64_t>()[i] =
+          inverse_map[input.data<scalar_t>()[i]];
```
I think you forgot to resize the tensor, so this will result in invalid memory accesses. Also, take the `.data` outside of the loop, it has a bunch of type checks inside IIRC and we don't want to do them at every step

theweiho replied:
Some lines above, I called `inverse_indices.resize_(input.sizes());`? Or are you referring to something else?

Done - refactored out `.data` calls.

apaszke commented on aten/src/ATen/native/Unique.cpp:
```diff
@@ -0,0 +1,58 @@
+  const Tensor& input = self.contiguous();
+  set_type<scalar_t> set(
+      input.data<scalar_t>(), input.data<scalar_t>() + input.numel());
+  Tensor output = input.type().tensor({static_cast<long long>(set.size())});
```
I think the cast should be to `int64_t`. It only happens to be an alias for `long long` in your compiler

theweiho replied:
Done

apaszke commented on test/test_torch.py:
```diff
@@ -5422,6 +5422,70 @@ def test_set_flush_denormal(self):
+            expected_unique.tolist(), sorted(x_unique.tolist()))
+        self.assertEqual(empty_inverse, x_inverse)
+
+        x_unique, x_inverse = x.unique(return_inverse=True)
```
If we want to be consistent with numpy we should only return the inverse if the user asked for it. Not a great design decision, but I don't think it's a good idea to diverge when we're using the same name

theweiho replied:
TensorFlow's unique op also returns the inverse (https://www.tensorflow.org/api_docs/python/tf/unique), and same for Caffe2's (https://caffe2.ai/docs/operators-catalogue.html#unique). I'm happy to change any arg name to something else if you prefer, but I don't think it's worth diverging from the design decision convention if consistency is your main concern.

theweiho committed:
Refactor out .data calls

colesbury commented on aten/src/ATen/native/cuda/Unique.cu:
```diff
@@ -0,0 +1,16 @@
+#include "ATen/ATen.h"
+
+#include <tuple>
+A
```
This looks like a stray character snuck in

goldsborough replied:
Is the CI disabled?

colesbury replied:
Yeah, I think @theweiho's account wasn't whitelisted

theweiho replied:
Fixed

colesbury commented:
@pytorchbot add to whitelist

ezyang commented:
@pytorchbot retest this please

ezyang commented:
@pytorchbot retest this please

ezyang commented:
@pytorchbot retest this please

theweiho committed:
Refactored to variable return length API with wrapper fn as opposed to returning a 0-length tensor, per off-line reviewer comments

theweiho committed:
Remove A

theweiho committed:
Don't use hidden torch._unique() in test

theweiho committed:
Fix documentations

ezyang commented on test/test_torch.py:
```diff
@@ -5422,6 +5422,67 @@ def test_set_flush_denormal(self):
+        x_unique, x_inverse = x.unique(return_inverse=True)
+        self.assertEqual(
+            expected_unique.tolist(), sorted(x_unique.tolist()))
+        self.assertEqual(expected_inverse.numel(), x_inverse.numel())
```
Isn't the inverse here dependent on the hashing strategy used in the implementation?

theweiho replied:
The inverse should always have the same number of elements (the same as the input), regardless of the hashing strategy...? Notice that I'm only comparing `numel()`

ezyang reviewed (dismissed):
I don't think the tests in the hashing case are correct.
